oak-segment-aws/src/main/java/org/apache/jackrabbit/oak/segment/aws/AwsContext.java

```diff
@@ -0,0 +1,441 @@
+
+    private final AmazonS3 s3;
+    private final String bucketName;
+    private final String rootDirectory;
+
+    private final AmazonDynamoDB ddb;
+    private final Table journalTable;
+    private final String lockTableName;
+
+    private RemoteStoreMonitor monitor;
+
+    private AwsContext(AmazonS3 s3, String bucketName, String rootDirectory, AmazonDynamoDB ddb,
+            String journalTableName, String lockTableName) {
+        this.s3 = s3;
+        this.bucketName = bucketName;
+        this.rootDirectory = rootDirectory.endsWith("/") ? rootDirectory : rootDirectory + "/";
+        this.ddb = ddb;
+        this.journalTable = new DynamoDB(ddb).getTable(journalTableName);
+        this.lockTableName = lockTableName;
+    }
+
+    private AwsContext(Configuration configuration) {
+        String region = configuration.region();
+        String rootDirectory = configuration.rootDirectory();
+        if (rootDirectory != null && rootDirectory.length() > 0 && rootDirectory.charAt(0) == '/') {
+            rootDirectory = rootDirectory.substring(1);
+        }
+
+        AWSCredentials credentials = configuration.sessionToken() == null || configuration.sessionToken().isEmpty()
+                ? new BasicAWSCredentials(configuration.accessKey(), configuration.secretKey())
+                : new BasicSessionCredentials(configuration.accessKey(), configuration.secretKey(),
+                        configuration.sessionToken());
+        AWSStaticCredentialsProvider credentialsProvider = new AWSStaticCredentialsProvider(credentials);
+
+        RequestHandler2 handler = new RequestHandler2() {
+            @Override
+            public void afterError(Request<?> request, Response<?> response, Exception e) {
+                process(request, response, e);
+            }
+
+            @Override
+            public void afterResponse(Request<?> request, Response<?> response) {
+                process(request, response, null);
+            }
+
+            private void process(Request<?> request, Response<?> response, Exception e) {
+                if (monitor != null) {
+                    TimingInfo timing = request.getAWSRequestMetrics().getTimingInfo();
+                    if (timing.isEndTimeKnown()) {
+                        long requestDuration = timing.getEndTimeNano() - timing.getStartTimeNano();
+                        monitor.requestDuration(requestDuration, TimeUnit.NANOSECONDS);
+                    }
+
+                    if (e == null) {
+                        monitor.requestCount();
+                    } else {
+                        monitor.requestError();
+                    }
+                }
+            }
+        };
+
+        AmazonS3 s3 = AmazonS3ClientBuilder.standard().withCredentials(credentialsProvider).withRegion(region)
+                .withRequestHandlers(handler).build();
+        AmazonDynamoDB ddb = AmazonDynamoDBClientBuilder.standard().withCredentials(credentialsProvider)
```

Comment:
Is it ok to re-use the previous hander here? Our request metrics will be skewed b/c frequent calls to update journal...

Reply:
this matches what we're doing in oak-segment-azure - every journal update is a blob append operation which results in a network call and that gets logged